### PR TITLE
Implemented findall(in(interval), x::AbstractRange), fixes #52

### LINKED ISSUE #52: Quickly find elements of vector covered by interval
In various settings, I need to efficiently find the indices of all elements of a sorted vector/range that belong to an interval, e.g. `within_interval(1:0.5:8, 2..4) -> 3:7`, whereas `within_interval(1:0.5:8, Interval{:closed,:open}(2,4) -> 3:6`. Would this be an interesting addition to IntervalSets.jl?

For ranges, one can do a little bit of algebra to find the indices [O(1) complexity], whereas for sorted vectors I guess bisection [O(N*log(N))] is the best one can do.

### PR DESCRIPTION
**Pull request by jagot**
----

#

### CODE REVIEW

**Comment by jagot** [comment]
I realized I wanted to add the following function as well:

```julia
Base.findall(interval_d::Base.Fix2{typeof(in),Interval{L,R,T}}, x::AbstractArray)  where {L,R,T} =
    findall(v -> v ∈ interval_d.x, x)
```

but this breaks an ambiguity test:
```
IntervalSets: Test Failed at /Users/jagot/.julia/dev/IntervalSets/test/runtests.jl:27
  Expression: isempty(detect_ambiguities(IntervalSets, Base, Core))  
   Evaluated:
   isempty(Tuple{Method,Method}[(findall(interval_d::Base.Fix2{typeof(in),Interval{L,R,T}},
   x::AbstractArray) where {L, R, T} in IntervalSets at
   /Users/jagot/.julia/dev/IntervalSets/src/findall.jl:81,
   findall(p::Function, S::SparseArrays.AbstractSparseMatrixCSC) in
   SparseArrays at
   /Applications/Julia-1.4.app/Contents/Resources/julia/share/julia/stdlib/v1.4/SparseArrays/src/sparsematrix.jl:1400),
   (findall(interval_d::Base.Fix2{typeof(in),Interval{L,R,T}},
   x::AbstractArray) where {L, R, T} in IntervalSets at
   /Users/jagot/.julia/dev/IntervalSets/src/findall.jl:81,
   findall(p::Base.Fix2{typeof(in),T} where T,
   x::SparseArrays.AbstractSparseMatrixCSC) in SparseArrays at
   /Applications/Julia-1.4.app/Contents/Resources/julia/share/julia/stdlib/v1.4/SparseArrays/src/sparsematrix.jl:1419),
   (findall(interval_d::Base.Fix2{typeof(in),Interval{L,R,T}},
   x::AbstractArray) where {L, R, T} in IntervalSets at
   /Users/jagot/.julia/dev/IntervalSets/src/findall.jl:81,
   findall(p::Base.Fix2{typeof(in),T} where T,
   x::SparseArrays.SparseVector{#s662,Ti} where #s662) where Ti in
   SparseArrays at
   /Applications/Julia-1.4.app/Contents/Resources/julia/share/julia/stdlib/v1.4/SparseArrays/src/sparsevector.jl:733),
   (findall(interval_d::Base.Fix2{typeof(in),Interval{L,R,T}},
   x::AbstractArray) where {L, R, T} in IntervalSets at
   /Users/jagot/.julia/dev/IntervalSets/src/findall.jl:81,
   findall(p::Function, x::SparseArrays.SparseVector{#s662,Ti} where
   #s662) where Ti in SparseArrays at
   /Applications/Julia-1.4.app/Contents/Resources/julia/share/julia/stdlib/v1.4/SparseArrays/src/sparsevector.jl:709)])   
Stacktrace:
 [1] macro expansion at /Users/jagot/.julia/dev/IntervalSets/test/runtests.jl:27 [inlined]
 [2] macro expansion at /Users/julia/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.4/Test/src/Test.jl:1113 [inlined]
 [3] top-level scope at /Users/jagot/.julia/dev/IntervalSets/test/runtests.jl:24
```

The function seems to work as intended, so I'm not sure if the test should be amended, or what else.

**Comment by dlfivefifty** [comment]
Just fix the ambiguity?

**Comment by mcabbott** [comment]
That would be nice to have work, I was a bit surprised the default didn't just work. Here's where it goes, BTW:
```
findall(pred::Fix2{typeof(in)}, x::Union{AbstractArray, Tuple}) = _findin(x, pred.x)

function _findin(a::Union{AbstractArray, Tuple}, b)
    ind  = Vector{eltype(keys(a))}()
    bset = Set(b)
    @inbounds for (i,ai) in pairs(a)
        ai in bset && push!(ind, i)
    end
    ind
end
```
It's a little tempting to write `Base.Set(i::Interval) = i`...

**Comment by jagot** [comment]
Does this look good now?

File: test/findall.jl [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+    end
+
+    @testset "Large intervals" begin
+        @test findall(in(4..Inf), 2:2:10) == 2:5
+        @test findall(in(4..1e20), 2:2:10) == 2:5
+        @test isempty(findall(in(-Inf..(-1e20)), 2:2:10))
+    end
+
+    @testset "Reverse intervals" begin
+        @testset "with reverse" begin
```
Comment (mcabbott) [inline test/findall.jl:70]: You could remove one level of nesting here. Otherwise looks good to me.

Comment (jagot) [reply test/findall.jl]: 👍

**Comment by jagot** [comment]
@mcabbott @dlfivefifty Is it fine if I merge this and release as 0.5.1?

File: test/findall.jl [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+
+    @testset "Open interval" begin
+        assert_in_interval(x, OpenInterval(0.2,0.4), 6:8)
+    end
+
+    @testset "Random intervals" begin
+        @testset "L=$L" for L=[:closed,:open]
+            @testset "R=$R" for R=[:closed,:open]
+                for i = 1:1 # 20
+                    interval = Interval{L,R}(minmax(rand(),rand())...)
```
Comment (dlfivefifty) [inline test/findall.jl:37]: Is the random number seeded somewhere? e.g. `Random.seed!(0)`

Comment (jagot) [reply test/findall.jl:37]: No, it is not. In fact, I am not sure what best practice is?

Comment (dlfivefifty) [reply test/findall.jl:37]: It's generally a good idea to seed the random number generator so that the tests are deterministic: otherwise, degenerate choices of floating point numbers may break the tests on other machines.

You can do this by adding Random to the test dependencies in (Project.toml), and do
```julia
using Random
Random.seed!(0)
```
where `0` can be an arbitrary number.

Comment (jagot) [reply test/findall.jl:45]: 👍

**Comment by jagot** [comment]
I found I had already covered the corner case above in my B-spline code (🤦). Will probably cook up some BigFloat tests, de-duplicate the tests I have now and re-add @mcabbott's OffsetArrays examples.

**Comment by jagot** [comment]
Good catch with the BigFloats, there were indeed some corner cases for reverse ranges, that I think I've fixed now.

**Comment by mcabbott** [comment]
Many thanks for getting this done. 

I just found another edge case, which I'm not really sure this ought to try to support, but anyway here it is: 
```julia
julia> using IntervalSets                                          

julia> using Dates                                                 

julia> findall(in(Interval(Date(2020, 1, 8), Date(2020, 1, 22))), Date(2020):Week(1):Date(2021))                                      
ERROR: MethodError: no method matching isless(::Week, ::Int64)
Closest candidates are:
  isless(::Missing, ::Any) at missing.jl:87
  isless(::DoubleFloats.DoubleFloat{T}, ::Integer) where T<:Union{Float16, Float32, Float64} at /Users/me/.julia/packages/DoubleFloats/8fO4c/src/type/compare.jl:135
  isless(::AbstractFloat, ::Real) at operators.jl:158
  ...
Stacktrace:
 [1] <(::Week, ::Int64) at ./operators.jl:268
 [2] findall(::Base.Fix2{typeof(in),Interval{:closed,:closed,Date}}, ::StepRange{Date,Week}) at /Users/me/.julia/packages/IntervalSets/D282g/src/findall.jl:43
```

**Comment by jagot** [comment]
I don't think we should _not_ support it, but I find myself unwilling to get into date semantics :) 
However, it _does_ seem as only some means of testing if a timespan is negative is required, this already works (to my astonishment):
```julia
julia> w = Week(1)
1 week

julia> cld(3w,2w)
2
```

**Comment by jagot** [comment]
@mcabbott I think this is fairly easy to fix, once https://github.com/JuliaLang/julia/issues/36099 is taken care of:

```julia
    a,b = if δx < zero(δx)
        rev = findall(in(interval), reverse(x))
        isempty(rev) && return rev

        a = (il+ir)-last(rev)
        b = (il+ir)-first(rev)

        a,b
    else
        lx, rx = first(x), last(x)
        l = max(leftendpoint(interval), lx-one(δx))
        r = min(rightendpoint(interval), rx+one(δx))

        (l > rx || r < lx) && return 1:0

        a = il + max(0, round(Int, cld(l-lx, δx)))
        a += (a ≤ ir && (x[a] == l && L == :open || x[a] < l))

        b = min(ir, round(Int, cld(r-lx, δx)) + il)
        b -= (b ≥ il && (x[b] == r && R == :open || x[b] > r))

        a,b
    end
```

**Comment by mcabbott** [comment]
Thanks for digging, I haven't got back to this... although I think I was guilty of introducing the `rx+1` in the first place!
